File: Intersection.cpp
#include<bits/stdc++.h>
using namespace std;


class Solution {
public:
    vector<int> intersection(vector<int>& nums1, vector<int>& nums2) {

         //sizes of the arrays 
        int n = nums1.size();
        int m = nums2.size();

        //sort the arrays
        sort(nums1.begin(),nums1.end());
        sort(nums2.begin(),nums2.end());


        vector<int>res;

        //declare two pointers 
        int i=0,j=0;
        while(i < n and j < m){
           // if the elements at i and j matches 
            if(nums1[i] == nums2[j]){
                if(res.size() == 0){
                     res.push_back(nums1[i]);
                     i++;
                     j++;
                }else{
                     if(res[res.size()-1] == nums1[i]){
                         i++;
                         j++;
                     }else{
                          res.push_back(nums1[i]);
                          i++;
                          j++;
                     }
                }
            }else{ // IF they don't match
                  if(nums1[i] < nums2[j]){
                       i++;
                  }else{
                       j++;
                  }
            }
        }
        return res;
    }
};

int main(){
        int n,m;
        cin >> n >> m;
        vector<int>nums1(n),nums2(m);
        for(int i=0;i<n;i++)cin >> nums1[i];
        for(int i=0;i<m;i++)cin >> nums2[i];
        Solution obj;
        vector<int>res = obj.intersection(nums1,nums2);
        for(auto it: res)
            cout << it << " ";
        

}

File: Intersection.py
from typing import List

class Solution:
    def intersection(self, nums1: List[int], nums2: List[int]) -> List[int]:
        # sizes of the arrays
        n = len(nums1)
        m = len(nums2)
        
        # sort the arrays
        nums1.sort()
        nums2.sort()

        res = []

        # declare two pointers
        i = 0
        j = 0

        while i < n and j < m:
            # if the elements at i and j match
            if nums1[i] == nums2[j]:
                if not res:
                    res.append(nums1[i])
                    i += 1
                    j += 1
                else:
                    if res[-1] == nums1[i]:
                        i += 1
                        j += 1
                    else:
                        res.append(nums1[i])
                        i += 1
                        j += 1
            else:  # if they don't match
                if nums1[i] < nums2[j]:
                    i += 1
                else:
                    j += 1

        return res

if __name__ == "__main__":
    n, m = map(int, input().split())
    nums1 = list(map(int, input().split()))
    nums2 = list(map(int, input().split()))
    obj = Solution()
    res = obj.intersection(nums1, nums2)
    for i in res:
        print(i, end=" ")
